functions/src/aqi-calculation/index.ts

```diff
@@ -52,36 +59,67 @@ exports.thingspeakToFirestore = functions
         thingspeakInfo
       );
 
-      const resolvedPath = readingsSubcollection(knownSensor.id);
-
-      // Only add data if not already present in database.
-      // This happens if a sensor is down, so only old data is returned.
+      const docId = knownSensor.id;
+      const resolvedPath = readingsSubcollection(docId);
       const readingsRef = firestore.collection(resolvedPath);
+
+      // If data is not already present in the database add
+      // it to the historical readings
+      let firestoreSafeReading = defaultPm25BufferElement;
       if (
         (
           await readingsRef
             .where('timestamp', '==', Timestamp.fromDate(reading.timestamp))
             .get()
         ).empty
       ) {
-        const firestoreSafeReading = {
+        // Update reading with values
+        firestoreSafeReading = {
           timestamp: Timestamp.fromDate(reading.timestamp),
           channelAPm25: reading.channelAPm25,
           channelBPm25: reading.channelBPm25,
           humidity: reading.humidity,
           latitude: reading.latitude,
           longitude: reading.longitude,
         };
+        // Add to historical readings
         await readingsRef.add(firestoreSafeReading);
       }
 
-      // Delays the loop so that we hopefully don't overload Thingspeak, avoiding
-      // our program from getting blocked.
-      // Allocates up to a minute of the two minute runtime for delaying
-      const oneMinuteInMilliseconds = 60000;
-      const delayBetweenSensors = oneMinuteInMilliseconds / sensorList.length;
-      await new Promise(resolve => setTimeout(resolve, delayBetweenSensors));
+      // Add readings to the PM 2.5 buffer
+      const sensorDocRef = firestore.collection('sensors').doc(knownSensor.id);
+      const status =
+        knownSensor.data().pm25BufferStatus ?? bufferStatus.DoesNotExist;
+      // If the buffer status is In Progress we don't update the buffer at all
+      if (status === bufferStatus.Exists) {
+        // If the buffer exists, update normally
+        let pm25BufferIndex = knownSensor.data().pm25BufferIndex;
+        const pm25Buffer = knownSensor.data().pm25Buffer;
+
+        pm25Buffer[pm25BufferIndex] = firestoreSafeReading;
+        /* eslint-disable-next-line no-magic-numbers */
+        pm25BufferIndex = (pm25BufferIndex + 1) % pm25Buffer.length;
+
+        await sensorDocRef.update({
+          pm25BufferIndex: pm25BufferIndex,
+          pm25Buffer: pm25Buffer,
+        });
+      } else if (status === bufferStatus.DoesNotExist || status === undefined) {
+        // If the buffer does not exist, populate it with default values so
+        // it can be updated in the future
+        await sensorDocRef.update({
+          pm25BufferStatus: bufferStatus.InProgress,
+        });
+        populateDefaultBuffer(false, docId);
```

Comment:
Add comment that this function updates the bufferStatus to Exists once completed

Reply:
Also add comment that this costs an extra write